src/gui/accountstate.cpp

```diff
@@ -305,12 +306,17 @@ void AccountState::slotInvalidCredentials()
     qCInfo(lcAccountState) << "Invalid credentials for" << _account->url().toString()
                            << "asking user";
 
-    if (account()->credentials()->ready())
+    _waitingForNewCredentials = true;
+    setState(AskingCredentials);
+
+    if (account()->credentials()->ready()) {
         account()->credentials()->invalidateToken();
+        if (auto creds = qobject_cast<HttpCredentials *>(account()->credentials())) {
+            if (creds->refreshAccessToken())
+                return;
```

Comment:
This is a bit weird, it breaks encapsulation.
I know how much HttpCredential is spaghetti, but did you have any other solution in mind that could avoid this?

Maybe something like moving `refreshAccessToken` down to `AbstractCredentials`, which would fall-back to call `askFromUser()` itself.